### PR TITLE
16.0 [2227][MIG] project_task_reply_stage

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[2227](https://www.quartile.co/web#id=2227&action=771&model=project.task&view_type=form&menu_id=505)
depends on: [project_task_stage_state](https://github.com/OCA/project/pull/1009)
This module is moved from qrtl-private to qrtl-custom repo because this module depends on AGPL-3 module (project_task_stage_state)

### CODE REVIEW

File: project_task_reply_stage/models/mail_message.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+from odoo import api, models
+
+
+class MailMessage(models.Model):
+    _inherit = "mail.message"
+
+    @api.model_create_multi
+    def create(self, values_list):
+        messages = super(MailMessage, self).create(values_list)
+        for res in messages:
```
Comment (yostashiro) [inline project_task_reply_stage/models/mail_message.py]:
```suggestion
        for message in messages:
```
`res` should be used for returned values/records from a method.  It's weird to use it in the manner you do.

File: project_task_reply_stage/models/mail_message.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                            int(
+                                self.env["ir.config_parameter"]
+                                .sudo()
+                                .get_param(
+                                    "project_task_reply_stage.customer_reply_stage_id"
+                                )
+                            )
+                        )
+                        if task.project_id in reply_stage.project_ids:
+                            task.sudo().write({"stage_id": reply_stage.id})
```
Comment (yostashiro) [inline project_task_reply_stage/models/mail_message.py]: This looks overly cumbersome and inefficient. For example, check on the model and getting reply_stage should be done outside the for loop (I do not think multiple models ever get involved in a create if I'm not mistaken).

Try to reduce the if ladder by getting out of the loop with `continue` when conditions are not met, for example.

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: project_task_reply_stage/models/mail_message.py (hunk 2 above)
Comment (AungKoKoLin1997) [reply project_task_reply_stage/models/mail_message.py]: I don't think we can check the model outside the for loop.When I reply mail for sale order and project task and then fetch the mail, I got two message in create method with each model.

File: project_task_reply_stage/models/mail_message.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                )
+            ):
+                continue
+            if task.project_id and task.project_id.customer_reply_stage_id:
+                task.sudo().write(
+                    {"stage_id": task.project_id.customer_reply_stage_id.id}
+                )
+            else:
+                if task.project_id in reply_stage.project_ids:
+                    task.sudo().write({"stage_id": reply_stage.id})
```
Comment (yostashiro) [inline project_task_reply_stage/models/mail_message.py]:
```suggestion
            if task.state == "draft":
                continue
            user = message.author_id.user_ids[:1]
            if not (user and user.has_group("base.group_user")):
                continue
            reply_stage = task.project_id.customer_reply_stage_id
            if reply_stage:
                task.sudo().write({"stage_id": reply_stage.id})
                continue
            reply_stage_id = int(
                self.env["ir.config_parameter"]
                .sudo()
                .get_param("project_task_reply_stage.customer_reply_stage_id")
            )
            if reply_stage_id in project.type_ids.ids:
                task.sudo().write({"stage_id": reply_stage_id})
```
Just made the code concise.  Please confirm that this is valid.

File: project_task_reply_stage/models/mail_message.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# Copyright 2022 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
```
Comment (yostashiro) [inline project_task_reply_stage/models/mail_message.py]:
```suggestion
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

**Comment by yostashiro** [comment]
@AungKoKoLin1997 Can you please go ahead and create a PR for project_task_stage_state in qrtl-oca and and fix the CI of this PR?

**Comment by AungKoKoLin1997** [comment]
> Can you please go ahead and create a PR for project_task_stage_state in qrtl-oca and and fix the CI of this PR?

@yostashiro I am waiting for this [PR](https://github.com/OCA/project/pull/1009) to be a good state and merged.

**Comment by yostashiro** [comment]
@AungKoKoLin1997 We do not need to wait for the OCA PR to be merged for this one, IMO.  The possible design change discussed won't affect us.  We do not need to rush to process this PR either, however I'd rather reduce the backlog where going ahead doesn't impose any visible risk.

**Review by yostashiro — APPROVED** [review]

**Pull request merged**